### ISSUE TITLE
streaming: potential conflict row id after fail-over or scaling

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#8892]
We use the system time of each compute node locally when generating `row_id` in `RowIdGenerator`, with the spin loop in case of the time goes backward. However, if there's fail-over or scaling, it's possible to generate conflict IDs on different machines.

One possible solution is to use the timestamp extracted from the `Barrier.epoch`, which has total order in the whole cluster. The problem is that 64-bit is not that enough for row id, and the QPS will be limited to a very small value, if we embed a single timestamp (millisecond) and vnodes into the 64-bit integer (as follows).

```
/// Current format:
///
/// | timestamp | vnode id | sequence |
/// |-----------|----------|----------|
/// |  41 bits  | 10 bits  | 12 bits  |
```

So what if we put a range of timestamps (ms) here? It's also problematic, as the epoch number is generated by the meta service without any feedback from the compute nodes. So if we eventually run out of all row ids in this epoch, we'll have nothing to do with the unassigned records, unless we implement the unaligned checkpoint for `RowIdGenExecutor`.

Per discussion with @st1page and @yezizp2012, a simpler approach could be similar to how we handle the watermark when scaling. The timestamp can grow locally and unlimitedly on each compute node, while it'll be persisted on each checkpoint. After scaling or fail-over, we'll initialize `RowIdGenerator` according to the last timestamp in the state table for each vnode, and sleep if necessary to avoid any conflict. #8866 can be resolved along with this design.

**Comment by fuyufjh** [comment]
> So what if we put a range of timestamps (ms) here? It's also problematic, as the epoch number is generated by the meta service without any feedback from the compute nodes. So if we eventually run out of all row ids in this epoch, there'll be nothing we can do about the unassigned records

In theory, you can decrease the precesion of `epoch` to seconds, so that `timestamp` can take fewer bits and `sequence` can take more bits. 🤣 BUT I don't like this solution. That would be a big change.

Another approach I can think up is, since we do pause and resume when reschedule the actors, as long as the duration of pause/resume is larger than the max difference between each CN's system time, it would be fine. Technically, those CN nodes's system time should not have second-level 
error because they are in the same data center.

**Comment by BugenZhao** [comment]
> In theory, you can decrease the precesion of `epoch` to seconds, so that `timestamp` can take fewer bits and `sequence` can take more bits. 🤣 BUT I don't like this solution. That would be a big change.

I agree. This will result in the barrier interval being limited to a granularity of seconds.

> as long as the duration of pause/resume is larger than the max difference between each CN's system time, it would be fine

Practically yes, but what I'm afraid of is the problem in some rare cases since currently we have no last resort to handle that. 🤣

**Comment by st1page** [comment]
Does this issue still exist 🥵 ?
